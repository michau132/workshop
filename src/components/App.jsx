import React, { Component } from 'react';
import { Link, Switch, Route, withRouter } from "react-router-dom";
import { AppBar, Toolbar, IconButton, Typography, Drawer, withStyles, Divider, List, ListItem, ListItemIcon, ListItemText } from "@material-ui/core";
import { MenuRounded, ChevronLeft, InboxOutlined } from "@material-ui/icons";

import routes from '../routes';
import './App.scss';
import getCurrentRoute from '../utils/getCurrentRoute';
import AppService from '../context/AppContext';

const drawerWidth = 300;

const CustomDrawer = withStyles(theme => ({
  paper: {
    width: drawerWidth,
    boxShadow: theme.shadows[8],
  },
}))(Drawer);

const CustomAppBar = withStyles((theme) => {
  // console.log(theme);
  return {
    root: {
      transition: theme.transitions.create([ 'width', 'margin', ], {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.leavingScreen,
      }),
    },
  };
})(AppBar);

class App extends Component {
  constructor() {
    super();
    this.state = {
      drawerOpen: true,
    };
    this.toggleDrawer = this.toggleDrawer.bind(this);
  }
 
  toggleDrawer() {
    this.setState(prevState => ({ drawerOpen: !prevState.drawerOpen, }));
  }
  render(){
    const { drawerOpen, } = this.state;
    const { title, } = getCurrentRoute(this.props.location);
    return (
      <div className="app">
        <AppService>
          <CustomAppBar 
            position="static"
            className={drawerOpen ? 'drawerOpen': ''}
          >
            <Toolbar>
              <IconButton edge="start" color="inherit" aria-label="menu" onClick={this.toggleDrawer}>
                <MenuRounded />
              </IconButton>
              <Typography variant="h6" >
                {title}
              </Typography>
            </Toolbar>
          </CustomAppBar>
          <CustomDrawer
            variant="persistent"
            anchor="left"
            open={drawerOpen}
          >
            <div className='iconClose'>
              <IconButton onClick={this.toggleDrawer} data-test='icon-button'>
                <ChevronLeft /> 
              </IconButton>
            </div>
            <Divider />
            <List>
              {routes
                .filter(el => !el.path.includes(':'))
                .map((route, index) => (
                  <Link to={route.path} key={index}>
                    <ListItem button>
                      <ListItemIcon>
                        <InboxOutlined />
                      </ListItemIcon>
                      <ListItemText>{route.title}</ListItemText>
                    </ListItem>
                  </Link>
                ))}
            </List>
          </CustomDrawer>
          <div className={`${drawerOpen ? 'drawerOpen': ''} mainContent`} >
            <Switch>
              {routes.map((route, index) => (
                <Route key={index} path={route.path} exact component={route.component} />
              ))}
            </Switch>
          </div>
        </AppService>
      </div>
    );
  }
}
export default withRouter(App);