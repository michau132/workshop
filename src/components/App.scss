
.iconClose {
  display: flex;
  justify-content: flex-end;
  padding: 5px 20px;
}

.drawerOpen, 
.drawerOpen.MuiPaper-root {
  width: calc(100% - 300px);
  margin-left: 300px;
}
.mainContent {
  padding: 20px 40px;
  max-width: calc(100% - 380px);
}