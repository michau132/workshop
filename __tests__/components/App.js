import React from 'react';
import { shallow } from "enzyme";
import App from "../../src/components/App";
import toJson from 'enzyme-to-json';

let cmp;
describe('<App /> component', () => {
  beforeEach(() => {
    cmp = shallow(<App.WrappedComponent />);
  });
  test('should render', () => {
    
    expect(cmp).toBeDefined();
  });
  
  test('should create snapshot', () => {
    const tree = toJson(cmp);
    expect(tree).toMatchSnapshot();

  });

  test('should update App state', () => {
    cmp.find('[data-test="icon-button"]').simulate('click');
    expect(cmp.state.drawerOpen).toBeFalsy();
  });
  
});
