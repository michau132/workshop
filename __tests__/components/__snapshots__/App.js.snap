// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> component should create snapshot 1`] = `
<div
  className="app"
>
  <AppService>
    <WithStyles(WithStyles(ForwardRef(AppBar)))
      className="drawerOpen"
      position="static"
    >
      <WithStyles(ForwardRef(Toolbar))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="menu"
          color="inherit"
          edge="start"
          onClick={[Function]}
        >
          <MenuRoundedIcon />
        </WithStyles(ForwardRef(IconButton))>
        <WithStyles(ForwardRef(Typography))
          variant="h6"
        />
      </WithStyles(ForwardRef(Toolbar))>
    </WithStyles(WithStyles(ForwardRef(AppBar)))>
    <WithStyles(WithStyles(ForwardRef(Drawer)))
      anchor="left"
      open={true}
      variant="persistent"
    >
      <div
        className="iconClose"
      >
        <WithStyles(ForwardRef(IconButton))
          data-test="icon-button"
          onClick={[Function]}
        >
          <ChevronLeftIcon />
        </WithStyles(ForwardRef(IconButton))>
      </div>
      <WithStyles(ForwardRef(Divider)) />
      <WithStyles(ForwardRef(List))>
        <Link
          key="0"
          to="/"
        >
          <WithStyles(ForwardRef(ListItem))
            button={true}
          >
            <WithStyles(ForwardRef(ListItemIcon))>
              <InboxOutlinedIcon />
            </WithStyles(ForwardRef(ListItemIcon))>
            <WithStyles(ForwardRef(ListItemText))>
              Workshop React
            </WithStyles(ForwardRef(ListItemText))>
          </WithStyles(ForwardRef(ListItem))>
        </Link>
        <Link
          key="1"
          to="/users"
        >
          <WithStyles(ForwardRef(ListItem))
            button={true}
          >
            <WithStyles(ForwardRef(ListItemIcon))>
              <InboxOutlinedIcon />
            </WithStyles(ForwardRef(ListItemIcon))>
            <WithStyles(ForwardRef(ListItemText))>
              User list
            </WithStyles(ForwardRef(ListItemText))>
          </WithStyles(ForwardRef(ListItem))>
        </Link>
      </WithStyles(ForwardRef(List))>
    </WithStyles(WithStyles(ForwardRef(Drawer)))>
    <div
      className="drawerOpen mainContent"
    >
      <Switch>
        <Route
          component={[Function]}
          exact={true}
          key="0"
          path="/"
        />
        <Route
          component={[Function]}
          exact={true}
          key="1"
          path="/users"
        />
        <Route
          component={[Function]}
          exact={true}
          key="2"
          path="/user/:id"
        />
      </Switch>
    </div>
  </AppService>
</div>
`;
